### PR TITLE
enhancement: Upgrade all VecBuffer sinks to allow setting `max_bytes`

### LINKED ISSUE #2917: Rework all the sinks that use `VecBuffer` to allow `batch.max_bytes`
The older `struct VecBuffer` is unable to count the byte size of events that are added to it. This prevents them from being able to enforce the `batch.max_bytes` limit. The newer `struct VecBuffer2` introduced in PR #2916 is able to track this size, but requires that the input type implements `trait Length`. For composite types, this requires a bit of thought and may not be possible without actually encoding the event.

Once all sinks are migrated over to the new `struct VecBuffer2`, it can be moved over the old name.

* [X] aws_cloudwatch_logs (PR #2916 )
* [ ] aws_kinesis_firehose
* [ ] aws_kinesis_streams
* [ ] loki
* [ ] BatchSink tests

### PR DESCRIPTION
**Pull request by bruceg**
This enhances all the remaining sinks that use `VecBuffer` to calculate the size of records being added to the buffer which enables them to set `batch.max_bytes`. This contains:

* [X] aws_kinesis_firehose
* [X] aws_kinesis_streams
* [X] loki
* [X] BatchSink tests

Particular attention should be paid to the Loki sink, as I have transformed the whole batching mechanism. I think it is mostly just a matter of moving the code from one place to another to be able to calculate lengths intelligently.

Closes #2917 

*Note:* This is based on top of #3147, so it will need to be rebased before it can be committed.

### CODE REVIEW

**Comment by bruceg** [comment]
Ok, #3147 is merged so this is ready for review.

**Comment by bruceg** [comment]
Oh, I also couldn't find a reference to the maximum batch size in bytes for the Loki sink. That needs to be addressed before this is ready to merge.

**Comment by binarylogic** [comment]
> Oh, I also couldn't find a reference to the maximum batch size in bytes for the Loki sink. That needs to be addressed before this is ready to merge.

@cyriltovena would you mind helping us out here? 🙏

**Comment by cyriltovena** [comment]
We don’t really have a max, we recommend to send at least every second or when the batch contains 100kb of logs.

So at least 100kb not if I answered the question?

**Comment by binarylogic** [comment]
Thanks, that helps 👍

File: src/sinks/util/buffer/loki.rs [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+            .use_size_as_events()?
+            .get_settings_or_default(defaults))
+    }
+
+    fn push(&mut self, mut item: Self::Input) -> PushResult<Self::Input> {
+        let labels_len = item
+            .labels
+            .iter()
+            .map(|label| label.0.len() + label.1.len() + LABEL_OVERHEAD)
+            .sum::<usize>()
```
Comment (fanatid) [inline src/sinks/util/buffer/loki.rs:87]:
```suggestion
            .fold(0, |total, label| total + label.0.len() + label.1.len() + LABEL_OVERHEAD)
```

**Review by fanatid — APPROVED** [review]

File: src/sinks/util/buffer/loki.rs (hunk 1 above)
Comment (bruceg) [reply src/sinks/util/buffer/loki.rs:87]: I don't think that will either help code generation or expressiveness.

**Review by binarylogic — APPROVED** [review]
Looks great, nice work.

**Review by lukesteensen — APPROVED** [review]

**Pull request merged**